### ISSUE TITLE
fromPromise bug with extendObservable in mobx4

### ISSUE THREAD

**Issue by rzcoder** [mobxjs/mobx-utils#119]
If try cast promise to observable 2 times, like:
```
import { fromPromise } from "mobx-utils";

let promise = Promise.resolve();
let observablePromise = fromPromise(promise);
observablePromise = fromPromise(promise);
```

causes

```
Error: [mobx] 'extendObservable' can only be used to introduce new properties. Use 'set' or 'decorate' instead. The property 'value' already exists on '[object Promise]'
```

**Comment by mweststrate** [comment]
It probably should be a no-op, but why do you cast it twice?

**Comment by rzcoder** [comment]
Some logic when need make external promise observable in few isolated components ¯\_(ツ)_/¯

**Comment by elado** [comment]
same here, use case is a function that returns the same promise, and computed property that wraps it with fromPromise.

```js
const api = {
  cachePromises: {},
  load(id) {
    if (!this.cachePromises[id]) this.cachePromises[id] = fetch(`/blah/${id}`)...
    return this.cachePromises[id]
  }
}

// store
class Store {
  @observable id = 'a'
  @computed get loadPromise() { return fromPromise(api.load(this.id)) }
}
```

the intention is when store.id changes it would load another one, but fromPromise mutates the [original promise](https://github.com/mobxjs/mobx-utils/blob/master/src/from-promise.ts#L77-L85).

So question is - should it be a no-op if it recognizes it has already been extended, or fromPromise should return a new promise?

**Comment by urugator** [comment]
Is the current behavior actually intended?
To me it seems that: 
```javascript
// https://github.com/mobxjs/mobx-utils/blob/master/src/from-promise.ts#L64
const promise = origPromise as any
origPromise.then(/**/)
```
should actually be
```javascript
const promise = origPromise.then(/**/) as any;
```
So that the `promise.then` is chained after the update of observables rather than running "concurrently" (due to JS sync nature it also runs after, but semantically it's concurrent)   

Also doc suggests different behavior:
> `fromPromise` takes a Promise and returns a new Promise wrapping the original one.

**Comment by mweststrate** [comment]
@urugator good point, but I noticed that attaching error handlers became a bit weird in that case, failing some unit tests (might have been an error on my side of course). Will re-investigate if it causes trouble somewhere in the future

Anyway, the original bug reported in this issue has been fixed in 4.0.1
